test/ASTGen/stmts.swift

```diff
@@ -33,3 +32,112 @@ func test1(e b: Bool) {
     print("FALSE")
   }
 }
+
+struct UnknownError: Error {}
+
+func testSwitch(arg1: Int, arg2: Int) throws {
+  LABEL: switch (arg1, arg2) {
+  case (1, 2):
+    print("")
+    fallthrough
+  case (1, _) where true, (4, 3):
+    break LABEL
+  case (var foo, let bar) where foo.isMultiple(of: 2):
+    defer {
+      print(foo)
+    }
+    foo += 1
+    print(bar)
+    break
+  default:
+    throw UnknownError()
+  }
+}
+
+func canThrow() throws {}
+
+func testDo() {
+  do {
+    try canThrow() 
+  } catch is UnknownError {
+    return 
+  } catch where error is UnknownError {
+    return 
+  } catch let error, {
+    _ = error
+    return 
+  } catch {
+    return
+  }
+
+  LOOP: do {
+    print("foo")
+    if true {
+      continue  LOOP
+    }
+  }
+}
+
+func testGuard(arg1: Int?, arg2: String) {
+  guard let arg1, var arg2First = arg2.first, arg1 > 2 else {
+    return
+  }
+  arg2First = "a"
+
+  print(arg1, arg2First)
+}
+
+func testFor(arg1: [Int?]) {
+  for _ in 0..<1 { }
+  for case let .some(elem) in arg1 where elem < 42 { print(elem) }
+  for var elem in [1,2,3] {
+    elem += 1
+    print(elem)
+  }
+}
+
+func testRepeat() {
+  repeat {
+    print(1)
+  } while true
+}
+
+func testRepeat(arg: [Int]) {
+  var iter = arg.makeIterator()
+  while let a = iter.next(), a > Int.random(in: 0..<10) {
+    print(a)
+  }
+}
+
+func testThen() {
+  let x: Int = if .random() {
+    then .zero
+  } else {
+    then 0
+  }
+}
+
+struct GenericTypeWithYields<T> {
+  var storedProperty: T?
+
+  var property: T {
+    _read {
+      yield storedProperty!
+    }
+    _modify {
+      yield &storedProperty!
+    }
+  }
+
+/* FXIME: yield(...) is parsed as an expression. ASTGen needs to onvert it to a statments.
```

Comment:
We should fix that in the new parser to parse `yield()` as a yield stmt, I think. Or do we need more context knowledge to decide whether `yield(…)` is a statement than we have in the new parser/

Reply:
Yeah, it's contextual. `yield` statements only appear in coroutine accessors (`_read` and `_modify`) https://github.com/apple/swift/blob/a6c98879abefd42d67aeeb3394784281061d1647/include/swift/Parse/Parser.h#L609-L613
I'm on the fence whether `SwiftParser` should recognize this context.

Reply:
Oh, that’s … an unfortunate choice in the old parser